Import data from text/copy&paste [SATURN-1036]

ehanna4 opened the pull request:
- Add ability for users to import data to the data table from text/copy&pasting data.
- Update the error messages in the modal for file and text data
- Fix small issue with key not being updated so that table will re-render automatically with a new import
- Added tool tip for upload button

ehanna4 committed:
Add ability to update data model with pasted in text

- Add ability for user to paste in and edit text to update the data model (previously users only had option to upload a .tsv file for updating the data model)
- TODO: needs code cleanup and discussion with Joy/UX

ehanna4 committed:
Fix Stylings of data import modal

Update the error message styling and some basic corrections to spacing and font weight

ehanna4 committed:
Update key to rerender table

key was not being updated by refreshKey appropriately, now it is so the data table will re-render automatically when new data is imported

ehanna4 committed:
Remove unused const

ehanna4 committed:
Clean up code in onChange for pasted text

Deduplicate code updating pasted text data and move to it's own function

ehanna4 commented:
![Confirm window ](https://user-images.githubusercontent.com/54322292/64704019-90633f00-d47b-11e9-83db-1d8d68e05813.png)
![Error Message invalid file](https://user-images.githubusercontent.com/54322292/64704020-90633f00-d47b-11e9-9532-f9db376f1e3c.png)
![Error Messge invalid pasted data](https://user-images.githubusercontent.com/54322292/64704021-90633f00-d47b-11e9-93dd-f7dfe9054222.png)
![ToolTip-abled](https://user-images.githubusercontent.com/54322292/64704022-90633f00-d47b-11e9-8154-c919a2462f0c.png)
![ToolTip-Disabled](https://user-images.githubusercontent.com/54322292/64704023-90633f00-d47b-11e9-8940-c8e07f72fb1d.png)
![Updated Modal look](https://user-images.githubusercontent.com/54322292/64704024-90633f00-d47b-11e9-8c27-f17a7c077f0d.png)
![Warning overwrite same data model](https://user-images.githubusercontent.com/54322292/64704025-90fbd580-d47b-11e9-823a-d4aeda15c708.png)

panentheos commented on src/libs/data-utils.js:
```diff
@@ -329,12 +341,70 @@ export const EntityUploader = class EntityUploader extends Component {
+            ...Utils.newTabLinkProps
+          }, ['Learn more ', icon('pop-out', { size: 12 })])
+        ]),
+        h(TextArea, {
```
UX question: Do we want to allow/encourage editing the pasted value here? Text areas don't provide a good experience for editing structured data, especially TSV (since it's hard to insert tabs), so I wonder if it would be better to disallow direct editing. We could still provide a preview area if that was helpful (and could even preview file contents as well).

panentheos commented on src/libs/data-utils.js:
```diff
@@ -329,12 +341,70 @@ export const EntityUploader = class EntityUploader extends Component {
           },
-          onClick: openUploader
+          onClick: pastedText ?
+            () => window.confirm(
```
Native alerts boxes like `confirm` block all JS execution while they're active, so they're generally not used on interactive sites. We'll want to find another way to handle this interaction.

panentheos commented on src/libs/data-utils.js:
```diff
@@ -329,12 +341,70 @@ export const EntityUploader = class EntityUploader extends Component {
+            style: { marginLeft: '1rem', verticalAlign: 'middle' },
+            href: 'https://software.broadinstitute.org/firecloud/documentation/article?id=10738',
+            ...Utils.newTabLinkProps
+          }, ['Learn more ', icon('pop-out', { size: 12 })])
```
There's some duplication of content between this case and the file upload case that would be good to factor out. TOL: Actually, in this case I wonder some UX tweaks would be a better option: Since the file and paste are mutually exclusive, we might be able to communicate that clearly by placing them side-by-side, and then we'd only need a single instance of the message that could apply to either input. Thoughts?

breilly2 replied:
I haven't studied the code yet, but I did look over the screenshots. It reminds me of the experience in [Insomnia](https://insomnia.rest/) of choosing a file as the request body or manually entering text, JSON, etc.

![Screen Shot 2019-09-12 at 9 25 38 AM](https://user-images.githubusercontent.com/19228696/64787864-61160600-d53f-11e9-90f2-4d7798080d23.png)

They do a similar thing where they warn that you'll lose the manually entered request body if you switch to using a file. Might be helpful for inspiration.

ehanna4 replied:
@panentheos I'm following up with UX about the vertical stacking and potentially changing it. I feel like putting both file and text import boxes inline horizontally might look squished in the modal though. Perhaps to cut down on the duplicate content of what's on the screen there could be a toggle and then only show relevant pieces? Then the user could switch to file import or text import and only the necessary import box and related text would show at one time?
@breilly2 I will download insomnia and take a look, thanks for the suggestion!

panentheos replied:
Yes, I like that idea too. Another potential benefit there is that if switching modes is an explicit action, it may be clear enough that we don't need an extra confirmation step.

ehanna4 replied:
Please see most recent changes - have removed the confirm and have moved the text vs paste imports to tabs

ehanna4 committed:
Add SimpleTabBar component and use in data import modal

- Add SimpleTabBar component and associated styling to common.js
- Update the data import modal to use this SimpleTabBar for switching between file and text imports

ehanna4 committed:
Remove unused imports

ehanna4 committed:
Remove confirm pop-up

remove the confirm pop-up and fix the error messages to only show when on the appropriate tab

breilly2 commented on src/libs/data-utils.js:
```diff
@@ -19,8 +20,8 @@ import * as Utils from 'src/libs/utils'
-  border: `1px solid ${colors.warning(0.85)}`,
-  backgroundColor: colors.warning(0.4),
+  border: '1px solid #DC8412',
+  backgroundColor: colors.warning(1),
```
Nit: Equivalent to `colors.warning()`.

Non-nit: This is now the same as the text color below. Looks like you're using `colors.light(.1)` for the text color of the warning you added. Should be able to put that here instead.

ehanna4 commented:
Closing pull request as work is lower priority and requires some changes. Will open a new pull request with new changes when able to work on it.
